### PR TITLE
Extend descriptor fast cache-hit path to optional inputs and resources

### PR DESCRIPTION
**Pull request by dgomezTT**
## Summary

Two framework changes that unlock the `BufferBinding` fast cache-hit path for migrated factories that currently fall through to the full descriptor rebuild on cache hits.

### 1. `emplace_runtime_args` accepts `nullptr` Buffer* as `0u` with no binding

Today, passing `nullptr` to `emplace_runtime_args(core, {Buffer*})` would try to register a binding for a null buffer. Factories migrating ops with optional inputs (e.g. `batch_norm`'s weight/bias, `attention_optimized_softmax`'s mask) currently work around this by using `runtime_args.emplace_back(buffer ? buffer->address() : 0u)` — which falls off the fast path because the slot is no longer registered.

After this change: `emplace_runtime_args(core, {maybe_null_buffer})` emits `0u` when the pointer is null and registers no binding for that slot. When the pointer is non-null it registers a binding as before. The fast path stays valid regardless of which optionals are present.

### 2. `collect_tensor_buffers` enumerates Tensor fields in `resource_t`

Resource tensors produced by `prepare_resources` (e.g. halo lookup tables) were not previously enumerated, so any factory that bound a kernel runtime arg to a resource Buffer* via `emplace_runtime_args` would fail in `resolve_bindings` with \"Buffer* not found in tensor_args/tensor_return_value enumeration.\"

After this change: `collect_tensor_buffers` also walks the `resources` struct via `ttsl::reflection::visit_object_of_type<Tensor>`. Resource Buffer* now resolves to a stable index, and the fast path can patch resource buffer addresses on cache hits like any other tensor.

## Test plan

- [x] Builds cleanly on the worktree
- [x] Smoke test: add + layer_norm + softmax repeated across iterations to exercise cache-hit path — passes
- [ ] Sanity: `runtime-sanity-tests.yaml`
- [ ] Sanity: `blackhole-post-commit.yaml`

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:dgomez/descriptor-framework-fast-path-extensions)

### CODE REVIEW

File: tt_metal/impl/program/program_descriptors.cpp [hunk 1]
```diff
@@ -203,8 +203,14 @@ static void emplace_runtime_args_impl(KernelDescriptor& kd, const CoreCoord& cor
...
-            kd.buffer_bindings.push_back({core, static_cast<uint32_t>(values.size()), *buf});
-            values.push_back((*buf)->address());
+            // nullptr Buffer* represents an absent optional tensor.  Emit 0u with no
+            // binding so the fast cache-hit path is not invalidated by optional inputs.
+            if (*buf == nullptr) {
+                values.push_back(0u);
+            } else {
+                kd.buffer_bindings.push_back({core, static_cast<uint32_t>(values.size()), *buf});
+                values.push_back((*buf)->address());
+            }
```
[the remark this answers is not in the document]
Comment (dgomezTT) [reply tt_metal/impl/program/program_descriptors.cpp:213]: Addressed in 9619868ffe5. Tests live in tests/tt_metal/tt_metal/api/test_descriptor_patching.cpp:

DescriptorPatching.EmplaceRuntimeArgs_NullBuffer_EmitsZero_NoBinding
DescriptorPatching.EmplaceCommonRuntimeArgs_NullBuffer_EmitsZero_NoBinding
Each one passes {u32, nullptr Buffer*, u32} and verifies the slot becomes 0u, no binding gets registered, and the call doesn't crash. All 10 DescriptorPatching.* tests green locally.

**Review by philei-tt — APPROVED** [review]

**Review by jbaumanTT — APPROVED** [review]
@akerteszTT looks like you might need to support optional Tensors in Metal 2.0

**Review by akerteszTT — APPROVED** [review]
Everything related to BufferBinding is a short-term shim pending Metal 2.0. Maybe mention in the PR description? Currently reads like 1st class infra.
